esp/esdllib/esdl_transformer2.cpp

```diff
@@ -895,7 +922,53 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, IPropertyTree *pt, const
             {
                 const char *tagname = child.queryInputName(ctx);
                 if (pt->hasProp(tagname)||child.hasDefaults())
-                    child.process(ctx, pt->queryPropTree(tagname), child.queryOutputName(ctx), &local, count);
+                {
+                    unsigned numChildren = pt->getCount(tagname);
+                    bool isEsdlList = false;
+                    IEsdlDefObject*  def = child.queryEsdlDefObject();
+                    if (def->getEsdlType() == EsdlTypeArray)
+                    {
+                        IEsdlDefArray* defArray = dynamic_cast<IEsdlDefArray*>(def);
+                        isEsdlList = defArray->checkIsEsdlList();
+                    }
+                    if (!isEsdlList)
+                    {
+                        if (numChildren < 2)
+                            child.process(ctx, pt->queryPropTree(tagname), child.queryOutputName(ctx), &local, count);
+                        else
+                        {
+                            Owned<IPropertyTreeIterator> iter = pt->getElements(tagname);
+                            ForEach(*iter)
+                                child.process(ctx, &iter->query(), child.queryOutputName(ctx), &local, count);
+                        }
+                    }
+                    else
+                    {
+                        Owned<IInterface> prevLocation = ctx.writer->saveLocation();
+                        ctx.writer->outputBeginArray(tagname);
+                        int curlen = ctx.writer->length();
+
+                        if (numChildren < 2)
+                            child.process(ctx, pt->queryPropTree(tagname), child.queryOutputName(ctx), &local, count);
+                        else
+                        {
+                            Owned<IPropertyTreeIterator> iter = pt->getElements(tagname);
+                            ForEach(*iter)
+                                child.process(ctx, &iter->query(), child.queryOutputName(ctx), &local, count);
+                        }
+                        if (ctx.writer->length()==curlen) //nothing was added... empty content remove opening tag
+                        {
+                            ///ctx.writer->outputEndArray(tagname);
```

Comment:
Have you tested this?  I'm wondering if the commented out "///ctx.writer->outputEndArray(tagname);" should be added back to get the state right?  or is the writer smart enough to figure out that the rewind removes the beginarray?